File: Lexer/Erlang/lexer.erl
-module(lexer).

-export([new/1, next_token/1, print_token/1, parse_while/1]).

-record(token, {type, literal = ""}).
-record(lexer, {input, curr = 0, peek = 0, ch}).

new(Input) ->
    L = #lexer{input = Input},
    advance(L).

parse_while(L) ->
    {T = #token{type = Type}, NewL} = next_token(L),
    case Type of
        eof ->
            ok;
        whitespace ->
            parse_while(NewL);
        _ ->
            print_token(T),
            parse_while(NewL)
    end.

advance(L = #lexer{input = Input, peek = P}) ->
    case P >= length(Input) of
        true ->
            L#lexer{ch = eof};
        false ->
            L#lexer{curr = P,
                    peek = P + 1,
                    ch = lists:nth(P + 1, Input)}
    end.

next_token(L = #lexer{ch = $*}) ->
    {#token{type = asterisk, literal = "*"}, advance(L)};
next_token(L = #lexer{ch = $/}) ->
    {#token{type = slash, literal = "/"}, advance(L)};
next_token(L = #lexer{ch = $-}) ->
    {#token{type = dash, literal = "-"}, advance(L)};
next_token(L = #lexer{ch = $+}) ->
    {#token{type = plus, literal = "+"}, advance(L)};
next_token(L = #lexer{ch = $=}) ->
    {#token{type = equal, literal = "="}, advance(L)};
next_token(L = #lexer{ch = $<}) ->
    {#token{type = lessthan, literal = "<"}, advance(L)};
next_token(L = #lexer{ch = $>}) ->
    {#token{type = greaterthan, literal = ">"}, advance(L)};
next_token(L = #lexer{ch = $;}) ->
    {#token{type = semicolon, literal = ";"}, advance(L)};
next_token(L = #lexer{ch = $:}) ->
    {#token{type = colon, literal = ":"}, advance(L)};
next_token(L = #lexer{ch = $(}) ->
    {#token{type = lparen, literal = "("}, advance(L)};
next_token(L = #lexer{ch = $)}) ->
    {#token{type = rparen, literal = ")"}, advance(L)};
next_token(L = #lexer{ch = ${}) ->
    {#token{type = lbrace, literal = "{"}, advance(L)};
next_token(L = #lexer{ch = $}}) ->
    {#token{type = rbrace, literal = "}"}, advance(L)};
% whitespace %
next_token(L = #lexer{ch = C}) when C == 32 orelse C >= 9 andalso C =< 13 ->
    {#token{type = whitespace}, advance(L)};
% alphabetic %
next_token(L = #lexer{ch = C})
    when C >= $a andalso C =< $z orelse C >= $A andalso C =< $Z orelse C == $_ ->
    {Ident, NewL} = read_identifier(L),
    Type =
        case Ident of
            "let" ->
                let_;
            "fn" ->
                function;
            "if" ->
                if_;
            "else" ->
                else_;
            "return" ->
                return;
            "true" ->
                boolean;
            "false" ->
                boolean;
            "bool" ->
                bool;
            "int" ->
                int;
            _ ->
                identifier
        end,
    {#token{type = Type, literal = Ident}, NewL};
% numeric %
next_token(L = #lexer{ch = C}) when C >= $0 andalso C =< $9 ->
    {Num, NewL} = read_number(L),
    {#token{type = number, literal = Num}, NewL};
% EOF %
next_token(L = #lexer{ch = eof}) ->
    {#token{type = eof}, advance(L)};
next_token(L = #lexer{ch = C}) ->
    {#token{type = illegal, literal = C}, advance(L)}.

% read_identifier %
read_identifier(L = #lexer{ch = C})
    when C >= $a andalso C =< $z orelse C >= $A andalso C =< $Z orelse C == $_ ->
    read_identifier(L, []).

read_identifier(L = #lexer{ch = C}, Acc)
    when C >= $a andalso C =< $z orelse C >= $A andalso C =< $Z orelse C == $_ ->
    NewL = advance(L),
    read_identifier(NewL, [C | Acc]);
read_identifier(L, Acc) ->
    {lists:reverse(Acc), L}.

% read_number %
read_number(L = #lexer{ch = C}) when C >= $0 andalso C =< $9 ->
    read_number(L, []).

read_number(L = #lexer{ch = C}, Acc) when C >= $0 andalso C =< $9 ->
    NewL = advance(L),
    read_number(NewL, [C | Acc]);
read_number(L, Acc) ->
    {lists:reverse(Acc), L}.

print_token(T = #token{literal = L}) ->
    case is_number(L) of
        true ->
            io:format("Token literal: \"~c\" type: ~s~n", [T#token.literal, T#token.type]);
        false ->
            io:format("Token literal: \"~s\" type: ~s~n", [T#token.literal, T#token.type])
    end.

File: Lexer/Elixir/lexer.exs
Code.require_file("token.exs")

defmodule Lexer do
  defstruct(input: "", curr: 0, peek: 0, char: nil)

  def new(input) do
    %Lexer{input: input}
    |> advance()
  end

  def parse_while(lexer) do
    {token = %Token{type: type}, new_lexer} = next_token(lexer)

    case type do
      :eof ->
        :ok

      :whitespace ->
        parse_while(new_lexer)

      _ ->
        print_token(token)
        parse_while(new_lexer)
    end
  end

  def advance(%Lexer{input: input, peek: p} = lexer) do
    if p >= String.length(input) do
      %Lexer{lexer | char: :eof}
    else
      %Lexer{
        lexer
        | curr: p,
          peek: p + 1,
          char: String.at(input, p) |> to_char_code()
      }
    end
  end

  def next_token(%Lexer{char: ?*} = lexer) do
    {%Token{literal: "*", type: :asterisk}, advance(lexer)}
  end

  def next_token(%Lexer{char: ?+} = lexer) do
    {%Token{literal: "+", type: :plus}, advance(lexer)}
  end

  def next_token(%Lexer{char: ?*} = lexer) do
    {%Token{literal: "*", type: :asterisk}, advance(lexer)}
  end

  def next_token(%Lexer{char: ?/} = lexer) do
    {%Token{literal: "/", type: :slash}, advance(lexer)}
  end

  def next_token(%Lexer{char: ?-} = lexer) do
    {%Token{literal: "-", type: :dash}, advance(lexer)}
  end

  def next_token(%Lexer{char: ?+} = lexer) do
    {%Token{literal: "+", type: :plus}, advance(lexer)}
  end

  def next_token(%Lexer{char: ?=} = lexer) do
    {%Token{literal: "=", type: :equal}, advance(lexer)}
  end

  def next_token(%Lexer{char: ?<} = lexer) do
    {%Token{literal: "<", type: :lessthan}, advance(lexer)}
  end

  def next_token(%Lexer{char: ?>} = lexer) do
    {%Token{literal: ">", type: :greaterthan}, advance(lexer)}
  end

  def next_token(%Lexer{char: ?;} = lexer) do
    {%Token{literal: ";", type: :semicolon}, advance(lexer)}
  end

  def next_token(%Lexer{char: ?:} = lexer) do
    {%Token{literal: ":", type: :colon}, advance(lexer)}
  end

  def next_token(%Lexer{char: ?(} = lexer) do
    {%Token{literal: "(", type: :lparen}, advance(lexer)}
  end

  def next_token(%Lexer{char: ?)} = lexer) do
    {%Token{literal: ")", type: :rparen}, advance(lexer)}
  end

  def next_token(%Lexer{char: ?{} = lexer) do
    {%Token{literal: "{", type: :lbrace}, advance(lexer)}
  end

  def next_token(%Lexer{char: ?}} = lexer) do
    {%Token{literal: "}", type: :rbrace}, advance(lexer)}
  end

  def next_token(%Lexer{char: c} = lexer) when c in 0..32 do
    {%Token{type: :whitespace}, advance(lexer)}
  end

  # alphabetic
  def next_token(%Lexer{char: c} = lexer) when c in ?a..?z or c in ?A..?Z or c == ?_ do
    {ident, newl} = read_identifier(lexer)
    {%Token{literal: ident, type: :identifier}, newl}
  end

  # numeric
  def next_token(%Lexer{char: c} = lexer) when c in ?0..?9 do
    {ident, newl} = read_number(lexer)
    {%Token{literal: ident, type: :number}, newl}
  end

  # EOF
  def next_token(%Lexer{char: :eof} = lexer) do
    {%Token{type: :eof}, advance(lexer)}
  end

  # all else (illegal)
  def next_token(%Lexer{char: c} = lexer) do
    {%Token{literal: c, type: :illegal}, advance(lexer)}
  end

  # identifier
  def read_identifier(%Lexer{char: c} = lexer) when c in ?a..?z or c in ?A..?Z or c == ?_ do
    read_identifier(lexer, [])
  end

  def read_identifier(%Lexer{char: c} = lexer, acc) when c in ?a..?z or c in ?A..?Z or c == ?_ do
    new_lexer = advance(lexer)
    read_identifier(new_lexer, [c | acc])
  end

  def read_identifier(lexer, acc) do
    {Enum.reverse(acc) |> IO.iodata_to_binary(), lexer}
  end

  # number
  def read_number(%Lexer{char: c} = lexer) when c in ?0..?9 do
    read_number(lexer, [])
  end

  def read_number(%Lexer{char: c} = lexer, acc) when c in ?0..?9 do
    new_lexer = advance(lexer)
    read_number(new_lexer, [c | acc])
  end

  def read_number(lexer, acc) do
    {Enum.reverse(acc), lexer}
  end

  def print_token(%Token{literal: literal, type: type}) do
    IO.puts("Token literal: \"#{literal}\" type: #{type}")
  end

  defp to_char_code(nil), do: nil
  defp to_char_code(<<c::utf8>>), do: c
end
